WalletWasabi/Wallets/FilterProcessor/BlockDownloadService.cs

```diff
@@ -0,0 +1,370 @@
+						if (!BlocksToDownload.TryDequeue(out Request? queuedRequest, out Priority? _))
+						{
+							throw new UnreachableException("Failed to dequeue block from the queue.");
+						}
+
+						Task newTask = HandleSingleBlockTaskAsync(queuedRequest, cancellationToken);
+						activeTasks.Add(newTask);
+					}
+				}
+
+				// It's still possible that there is no task because a reorg might have occurred.
+				if (activeTasks.Count == 0)
+				{
+					continue;
+				}
+
+				if (activeTasks.Count == MaximumParallelTasks)
+				{
+					Task task = await Task.WhenAny(activeTasks).ConfigureAwait(false);
+					activeTasks.Remove(task);
+				}
+			}
+		}
+		catch (OperationCanceledException)
+		{
+			Logger.LogDebug("Block-downloading service execution was stopped.");
+		}
+		catch (Exception ex)
+		{
+			Logger.LogError(ex);
+			throw;
+		}
+		finally
+		{
+			// Mark everything as cancelled because the service is shutting down (either gracefully or forcibly).
+			lock (Lock)
+			{
+				while (BlocksToDownload.TryDequeue(out Request? request, out _))
+				{
+					_ = request.Tcs.TrySetResult(CanceledResult.Instance);
+				}
+			}
+		}
+	}
+
+	private async Task HandleSingleBlockTaskAsync(Request request, CancellationToken cancellationToken)
+	{
+		RequestResponse response = await DownloadSingleBlockAsync(request, cancellationToken).ConfigureAwait(false);
+
+		if (response.Result is SuccessResult)
+		{
+			_ = request.Tcs.TrySetResult(response.Result);
+		}
+		else
+		{
+			Logger.LogDebug($"Attempt to download block {request.BlockHash} (height: {request.Priority.BlockHeight}) failed.");
+
+			// The block might have been removed concurrently if a reorg occurred.
+			if (!request.Tcs.TrySetResult(response.Result))
+			{
+				IResult opResult = await request.Tcs.Task.ConfigureAwait(false);
+				Logger.LogDebug($"Failed to set result for block '{request.BlockHash}' (height: {request.Priority.BlockHeight}). Result is already: {opResult}");
+			}
+		}
+	}
+
+	/// <summary>
+	/// Downloads a single block from a block source, or gets the block from the file-system cache if available.
+	/// </summary>
+	private async Task<RequestResponse> DownloadSingleBlockAsync(Request request, CancellationToken cancellationToken)
+	{
+		Logger.LogTrace($"Trying to download {request.BlockHash} (height: {request.Priority.BlockHeight}).");
+
+		try
+		{
+			// Try get the block from the file-system storage.
+			Block? block = await FileSystemBlockRepository.TryGetAsync(request.BlockHash, cancellationToken).ConfigureAwait(false);
+			if (block is not null)
+			{
+				return new RequestResponse(request, new SuccessResult(block, EmptySourceData.FileSystemCache));
+			}
+
+			SuccessResult? successResult = null;
+			ISourceData? failureSourceData = null;
+
+			if (request.SourceRequest is FullNodeSourceRequest)
+			{
+				if (TrustedFullNodeBlockProviders.Length == 0)
+				{
+					return new RequestResponse(request, NoSuchProviderResult.Instance);
```

Comment:
Interesting. The plan is to introduce some kind of new behavior? What happens if I set my local node as source and it cannot provides? We are not falling back to random nodes anymore?